services/webtheme/webtheme.go

```diff
@@ -28,48 +28,23 @@ const (
 )
 
 type ThemeMetaInfo struct {
-	FileName     string
-	InternalName string
-	DisplayName  string
+	FileName           string
+	InternalName       string
+	DisplayName        string
+	IsDarkTheme        bool
 }
 
+// extract CSS vars from CSS, taking the last occurence in a file to support combined themes like "auto"
 func parseThemeMetaInfoToMap(cssContent string) map[string]string {
-	/*
-		The theme meta info is stored in the CSS file's variables of `gitea-theme-meta-info` element,
-		which is a privately defined and is only used by backend to extract the meta info.
-		Not using ":root" because it is difficult to parse various ":root" blocks when importing other files,
-		it is difficult to control the overriding, and it's difficult to avoid user's customized overridden styles.
-	*/
-	metaInfoContent := cssContent
-	if pos := strings.LastIndex(metaInfoContent, "gitea-theme-meta-info"); pos >= 0 {
-		metaInfoContent = metaInfoContent[pos:]
-	}
-
-	reMetaInfoItem := `
-(
-\s*(--[-\w]+)
-\s*:
-\s*("(\\"|[^"])*")
-\s*;
-\s*
-)
-`
-	reMetaInfoItem = strings.ReplaceAll(reMetaInfoItem, "\n", "")
-	reMetaInfoBlock := `\bgitea-theme-meta-info\s*\{(` + reMetaInfoItem + `+)\}`
-	re := regexp.MustCompile(reMetaInfoBlock)
-	matchedMetaInfoBlock := re.FindAllStringSubmatch(metaInfoContent, -1)
-	if len(matchedMetaInfoBlock) == 0 {
-		return nil
-	}
-	re = regexp.MustCompile(strings.ReplaceAll(reMetaInfoItem, "\n", ""))
-	matchedItems := re.FindAllStringSubmatch(matchedMetaInfoBlock[0][1], -1)
 	m := map[string]string{}
-	for _, item := range matchedItems {
-		v := item[3]
-		v = strings.TrimPrefix(v, "\"")
-		v = strings.TrimSuffix(v, "\"")
-		v = strings.ReplaceAll(v, `\"`, `"`)
-		m[item[2]] = v
+
+	for _, v := range []string{"--theme-display-name", "--is-dark-theme"} {
+		re := regexp.MustCompile(v + `\s?:\s?["']?([^"';]+)["';]`)
+		matches := re.FindAllStringSubmatch(cssContent, -1)
+		numMatches := len(matches)
+		if numMatches > 0 {
+			m[v] = matches[numMatches - 1][1]
+		}
```

Comment:
1. Speed is irrelevant if it's only done once at startup and I don't imagine this taking more than a couple of milliseconds.
2. It's necessary for some use cases if you check frontend. It does resolve correctly in the auto theme as well because only one `:root`is active there.

Reply:
> It's necessary for some use cases if you check frontend. It does resolve correctly in the auto theme because only one `:root`is active there.

So my suggestion is to just keep the `--is-dark-theme` in the `:root` and let the theme describe itself by `mode` in our private block. They could be irrelevant. Let meta info block do backend work, `:root` do frontend work. Or it is also quite simple and light to make JS code get the variables in our meta info block, also "I don't imagine reading variables from our own private block taking more than a couple of milliseconds."

Reply:
> get the variables in our meta info block

That's actually not so easy because the normal way of `getComputedStyle` does not work on a non-matching element. So likely one has to parse `document.styleSheets` in JS which is a major pain in the ass. If we can make the private block match on something like `:root.gitea` I'd be more inclined to accept it.

Reply:
> That's actually not so easy because the normal way of `getComputedStyle` does not work on a non-matching element.

As simple as: `let el=document.createElement('gitea-theme-meta-info'); document.body.append(el);  window.getComputedStyle(el).getPropertyValue('--theme-display-name')` 

Or use `.gitea-theme-meta-info` and add the class to `<body class="gitea-theme-meta-info">` then read from `document.body`

Reply:
OK, actually, we should still separate the backend `--theme-scheme-mode` and frontend `--is-dark-theme`. Think about this case:

We we have only one variable:

```
@import light
@import dark

...
--theme-scheme-mode: auto 

```

The JS code could only get `--theme-scheme-mode == auto`, no matter light or dark scheme is applied.


So, `--is-dark-theme` could only exist in `:root` and should respect browser's scheme.

Meanwhile , `--theme-scheme-mode: auto ` is for backend only, and it shouldn't inherit.

Reply:
In short: JS code should never use the variables in `gitea-theme-meta-info`, because they are affected by the current scheme in auto them. So `gitea-theme-meta-info` should be backend-only and non-inheriting.

Reply:
> The JS code could only get --theme-scheme-mode == auto, no matter light or dark scheme is applied.

That's not really a supported usage, but essentially webpack will compile the auto theme to this:

````css
@media (prefers-color-scheme: dark) {
  :root {--is-dark-theme: true}
}
@media (prefers-color-scheme: light) {
  :root {--is-dark-theme: false}
}
````
The server can not know the user's preference, only the client will know, and must know for example to switch the monaco theme to dark, so this variable must stay inside the media blocks to work.

Reply:
I think the current approach can work as long as we document that we take the last occurence of the variable anywhere in the CSS.

Reply:
No, it is not right. Maybe there is some misunderstanding, but I am sure my approach is right. Unless there is any fundamental problem, I insist my solution.

Reply:
PS: I can work on docs if you approve the current method.

Reply:
The correct solution:

```
@media (prefers-color-scheme: dark) {
  gitea-theme-meta-info { --theme-scheme-mode: dark }
  :root {--is-dark-theme: true}
}
@media (prefers-color-scheme: light) {
  gitea-theme-meta-info { --theme-scheme-mode: light }
  :root {--is-dark-theme: false}
}

gitea-theme-meta-info { --theme-scheme-mode: auto }
```

Reply:
> PS: I can work on docs if you approve the current method.

current method is not right

Reply:
I still don't like that this selector matches nothing, there must be a better way.

Reply:
> I still don't like that this selector matches nothing, there must be a better way. 

It seems that it is just "you don't like". Does it really cause any problem?

> A strict linting tool that checks selector use would flag this an error.

All lint tools have the ability to allow pre-defined selectors. Just like you have defined many web components.

Again and last sentence for this problem: I believe my method is good enough and it is the right thing to do. Unless there is any fundamental problem, I insist my method.

Reply:
Both variants work fine but yours is more unergonomic both from user and js perspective. Maybe we should let others weight in, I will summarize below.

Reply:
Please propose a workable sample first:

1. Backend code could know the theme's mode: light, dark, auto (to display in the list).
2. JS code could know the whether the theme is light/dark (no auto), to decide how to apply colors.

Reply:
And I believe mine is more ergonomic both from user because there is a clear block to write various meta info. And, as I said many times, JS never needs these meta info.

Reply:
Auto theme detection will work fine as long as the var ist the last occurence in the compiled theme with current approach this wil work:

```
@import light
@import dark

--theme-scheme-mode: auto 
```

Reply:
More specifically, this will work for all use cases:

```css
@media (prefers-color-scheme: dark) {
  :root {
    --is-dark-theme: true;
  }
}

@media (prefers-color-scheme: light) {
  :root {
    --is-dark-theme: false;
  }
}

:root {
  --theme-scheme: auto;
}
```

Reply:
So  it still needs separate variables. And the dark/light are imported, so the full content should be like this.

Since you also introduce a new variable, why not separate `--theme-*` into a separate private block to make them more ergonomic?


```
@media (prefers-color-scheme: dark) {
  :root {
    --theme-display-name: "My Dark Theme";
    --theme-color-scheme: dark;
    --is-dark-theme: true;
  }
}

@media (prefers-color-scheme: light) {
  :root {
    --theme-display-name: "My Light Theme";
    --theme-color-scheme: light;
    --is-dark-theme: false;
  }
}

:root {
  --theme-display-name: "My Auto Theme";
  --theme-color-scheme: auto;
}
```

Reply:
And there is still a problem, inheriting.

As a template system, the variables should not inherit. See the example above, if we need to support "author/website" variables, then the `--theme-website` shouldn't inherit if it is not provided. For example:


```
@media (prefers-color-scheme: dark) {
  :root {
    --theme-display-name: "My Dark Theme";
    --theme-color-scheme: dark;
    --theme-website: gitea.com
    --is-dark-theme: true;
  }
}

@media (prefers-color-scheme: light) {
  :root {
    --theme-display-name: "My Light Theme";
    --theme-color-scheme: light;
    --theme-website: gitea.com
    --is-dark-theme: false;
  }
}

:root {
  --theme-display-name: "A 3rd theme";
  !!!!! it should NOT inherit the website !!!!!! it is not from gitea.com
}
```

Reply:
Problem is that with how CSS works, the last selector with same specificity wins, so the vars above would be overridden by the last selector.

Reply:
See https://jsfiddle.net/silverwind/73um04p8/ that showcases the problem.

Reply:
> And there is still a problem, inheriting.

Backend parser could just ignore anything inside `@media` blocks. But I think you are over-exaggerating a problem that does not exist.

Reply:
`gitea-theme-meta-info` is designed for backend-only, it is never used by CSS or JS. I don't know why you insist to use it in CSS/JS.

If frontend needs to use the template name, it should be provided by backend data. Do not use `gitea-theme-meta-info` in frontend directly.

Reply:
A `-theme-color-scheme` would be useful for frontend to set a `vscode-auto` theme for example. This would actually also solve a issue where we set a vscode theme only on page load, but if the user changes their OS color mode, vscode does not switch with it.

Reply:
> A `--theme-color-scheme` would be useful for frontend to set a `vscode-auto` theme for example.

It could done by something like `<tag data-theme-color-scheme="...">`, it is more stable and no CSS trick. Why CSS access is a must? Backend provided data could handle various cases better including there is no `theme-color-scheme` defined, and we do not need to (and shouldn't) tight our theme scheme to vscode theme scheme, there should be some wrapping or converting for better maintainability.

Reply:
CSS access is not a must for me but nice to have. And if the themes work correctly with CSS precedence, we don't block ourselves in case a need arises in the future, for example to display values in CSS pseudos like `content: var(--varname)`.